pkg/kubelet/kubelet_pods.go

```diff
@@ -968,19 +969,32 @@ func (kl *Kubelet) podResourcesAreReclaimed(pod *v1.Pod) bool {
 	return kl.PodResourcesAreReclaimed(pod, status)
 }
 
-// filterOutTerminatedPods returns pods that are not in a terminal phase
+// filterOutInactivePods returns pods that are not in a terminal phase
 // or are known to be fully terminated. This method should only be used
 // when the set of pods being filtered is upstream of the pod worker, i.e.
 // the pods the pod manager is aware of.
-func (kl *Kubelet) filterOutTerminatedPods(pods []*v1.Pod) []*v1.Pod {
+func (kl *Kubelet) filterOutInactivePods(pods []*v1.Pod) []*v1.Pod {
 	filteredPods := make([]*v1.Pod, 0, len(pods))
 	for _, p := range pods {
+		// if a pod is fully terminated by UID, it should be excluded from the
+		// list of pods
 		if kl.podWorkers.IsPodKnownTerminated(p.UID) {
 			continue
 		}
-		if p.Status.Phase == v1.PodSucceeded || p.Status.Phase == v1.PodFailed {
+
+		// terminal pods are considered inactive UNLESS they are actively terminating
+		isTerminal := p.Status.Phase == v1.PodSucceeded || p.Status.Phase == v1.PodFailed
+		if !isTerminal {
+			// a pod that has been marked terminal within the Kubelet is considered
+			// inactive (may have been rejected by Kubelet admision)
+			if status, ok := kl.statusManager.GetPodStatus(p.UID); ok {
+				isTerminal = status.Phase == v1.PodSucceeded || status.Phase == v1.PodFailed
+			}
```

Comment:
Just to confirm my understanding, the idea is that isTerminal when reading the pod phase i.e.

```
isTerminal := p.Status.Phase == v1.PodSucceeded || p.Status.Phase == v1.PodFailed
```

vs when reading from the statusManager

```
if status, ok := kl.statusManager.GetPodStatus(p.UID); ok {
				isTerminal = status.Phase == v1.PodSucceeded || status.Phase == v1.PodFailed
			}
```

would differ when kubelet would reject the pod during pod admission. In that case, only status manager would see the pod in terminal phase? Is that correct and why these two checks are required?

Reply:
Think about this as "how stale is the status we're getting as input".  If a pod on the apiserver is Running, and then the Kubelet decides to reject it, it takes some non-zero time for Kubelet to write the status back to apiserver so that it shows up in the watch and then comes back down to the Kubelet.  For that amount of time, "status" is authoritative and so the Kubelet knows that the pod is rejected but the apiserver doesn't.  If you restarted the Kubelet right after the decision was made but before the apiserver got the write, the new kubelet has no history other than what is on the apiserver, so it's going to start trying to start the pod again.

Conversely, if a control plane controller says "hey this failed because <insert arbitrary reason for a controller to mark a pod as Failed>" the Kubelet should honor it (which is the first check) and it overrides the local status (for instance, if local Kubelet saw the pod was "Succeeded" at the same time a control plane controller marked it "Failed", the correct outcome is "Failed", because both of those are terminal states for the phase state machine and first one wins).

The last wrinkle is a static pod.  A static pod has no other source of truth than a file on the local disk, so Kubelet considers it "always running".

Reply:
Thank a lot for the explanation, it's much more clear now. I think adding some comment to summarize your comment here would be really helpful. Maybe something along the lines of: 

```
// Consider pods as terminal if they are in succeeded of failed phase; terminal pods are considered inactive UNLESS they are actively terminating
// Pods can be put seen in terminal state by two sources which may not necessarily be in sync with each other: API server and local kubelet status manager. First, check pod phase as seen from api-server and also check local kubelet status manager (which may have a more up to date phase state) which should take precedence. 
```